File: src/main/java/io/github/emcw/utils/http/OfficialAPI.java
package io.github.emcw.utils.http;

import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import lombok.SneakyThrows;
import org.jetbrains.annotations.Nullable;

import java.util.concurrent.CompletableFuture;

public class OfficialAPI {
    public static String DOMAIN = "https://api.earthmc.net";

    public static class V3 {
        static final String BASE_ENDPOINT = DOMAIN + "/v3/aurora/";

        public static @Nullable JsonObject serverInfo() {
            // Server info doesn't have its own endpoint, it just lives at the base.
            JsonElement res = JSONRequest.sendGet(BASE_ENDPOINT);
            if (res == null) {
                return null;
            }

            return res.getAsJsonObject();
        }
    }

//    private @Nullable JsonObject townyData(@NotNull String endpoint) {
//        String fullUrl = OAPI_DOMAIN + (endpoint.startsWith("/") ? endpoint.substring(1) : endpoint);
//        CompletableFuture<JsonObject> data = get(fullUrl);
//
//        try {
//            return data.join().getAsJsonObject();
//        }
//        catch (Exception e) {
//            System.out.println(
//                "Error fetching OAPI data from " + fullUrl +
//                "\nReceived response may be incorrectly formatted."
//            );
//        }
//
//        return null;
//    }
}

File: src/main/kotlin/io/github/emcw/utils/http/OfficialAPI.kt
package io.github.emcw.utils.http

import com.google.gson.JsonObject
import org.jetbrains.annotations.Contract
import java.util.concurrent.*

class OfficialAPI {
    private fun townyData(endpoint: String): JsonObject? {
        val fullUrl = OAPI_DOMAIN + if (endpoint.startsWith("/")) endpoint.substring(1) else endpoint
        val data = Companion[fullUrl]
        try {
            return data.join().asJsonObject
        } catch (e: Exception) {
            println(
                """
                    Error fetching OAPI data from $fullUrl
                    Received response may be incorrectly formatted.
                    """.trimIndent()
            )
        }
        return null
    }

    fun serverInfo(): JsonObject? {
        // TODO: Proper deserialization of API response. (Using records?)
        return townyData("")

        //return null;
    }

    protected class Towns
    companion object {
        const val OAPI_DOMAIN = "https://api.earthmc.net/v2/aurora/"
        @Contract("_, -> new")
        private operator fun get(url: String): CompletableFuture<JsonObject> {
            return CompletableFuture.supplyAsync {
                try {
                    return@supplyAsync Request.send<JsonObject>(url)
                } catch (e: Exception) {
                    println(
                        """
    Exception occurred!
    ${e.message}
    """.trimIndent()
                    )
                    return@supplyAsync null
                }
            }
        }
    }
}